### ISSUE TITLE
Multiaccount OAuth2 - basic

### ISSUE THREAD

**Issue by jesmrec** [owncloud/android#2095]
### Steps to reproduce
1. Add a basic auth account
2. Add a OAuth2 account 
3. Switch to account in 1.

### Current behaviour
Basic auth account is redirected to login view

### Expected behaviour
Swiching of accounts correct
 
Android app 2.5.0
Samsung Galaxy S7 v7
Nexus 6P v7.1
Server v10.0

Stack trace:

01-17 12:31:28.175 32421-768/com.owncloud.android.debug E/HttpMethodDirector: bearer authorization challenge expected, but not found
                                                                              org.apache.commons.httpclient.auth.AuthenticationException: bearer authorization challenge expected, but not found
                                                                                  at org.apache.commons.httpclient.auth.AuthChallengeProcessor.processChallenge(AuthChallengeProcessor.java:159)
                                                                                  at org.apache.commons.httpclient.HttpMethodDirector.processWWWAuthChallenge(HttpMethodDirector.java:694)
                                                                                  at org.apache.commons.httpclient.HttpMethodDirector.processAuthenticationResponse(HttpMethodDirector.java:668)
                                                                                  at org.apache.commons.httpclient.HttpMethodDirector.executeMethod(HttpMethodDirector.java:193)
                                                                                  at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:397)
                                                                                  at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:323)
                                                                                  at com.owncloud.android.lib.common.OwnCloudClient.executeMethod(OwnCloudClient.java:234)
                                                                                  at com.owncloud.android.lib.resources.files.ReadRemoteFolderOperation.run(ReadRemoteFolderOperation.java:79)
                                                                                  at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:266)
                                                                                  at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:142)
                                                                                  at com.owncloud.android.operations.SynchronizeFolderOperation.fetchRemoteFolder(SynchronizeFolderOperation.java:253)
                                                                                  at com.owncloud.android.operations.SynchronizeFolderOperation.run(SynchronizeFolderOperation.java:206)
                                                                                  at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:266)
                                                                                  at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:142)
                                                                                  at com.owncloud.android.operations.common.SyncOperation.execute(SyncOperation.java:91)
                                                                                  at com.owncloud.android.operations.RefreshFolderOperation.run(RefreshFolderOperation.java:143)
                                                                                  at com.owncloud.android.lib.common.operations.RemoteOperation.runOperation(RemoteOperation.java:266)
                                                                                  at com.owncloud.android.lib.common.operations.RemoteOperation.run(RemoteOperation.java:232)
                                                                                  at java.lang.Thread.run(Thread.java:762)
01-17 12:31:28.212 3705-4896/? I/AccountManagerService: the accounts changed, sending broadcast of android.accounts.LOGIN_ACCOUNTS_CHANGED
01-17 12:31:28.218 32421-768/com.owncloud.android.debug E/ReadRemoteFolderOperation: Synchronized /: Operation finished with HTTP status code 401 (fail)
01-17 12:31:28.221 32421-768/com.owncloud.android.debug E/SynchronizeFolderOperation: Checked user1@docker.oc.solidgear.es:11306/ : Operation finished with HTTP status code 401 (fail)
01-17 12:31:28.224 32421-768/com.owncloud.android.debug D/RefreshFolderOperation: Send broadcast com.owncloud.android.operations.RefreshFolderOperation.EVENT_SINGLE_FOLDER_CONTENTS_SYNCED

**Comment by davigonz** [comment]
Well, I've been researching a bit about this issue and this is what I've found.

Our network library, Apache HTTP Client 3.x, **supports three different types of http authentication** schemes: Basic, Digest and NTLM, so to support SAML and OAuth, we introduced a **new scheme called Bearer**.

All the requests we perform to the server (once we are logged in) includes the Authorization header, even before the server responses with unauthorized response, since we are using **preemptive authentication** to avoid the overhead of making the connection for each request.

So, this is what we are currently doing to enable preemptive authentication in our library:

_Taking into account that **PREEMPTIVE_AUTH_SCHEME** is a constant in our library which value is "basic"._

- Basic (1):

`AuthPolicy.registerAuthScheme(AuthState.PREEMPTIVE_AUTH_SCHEME, BasicAuthScheme.class);`

`client.getParams().setAuthenticationPreemptive(true);`

- Bearer (2):

`AuthPolicy.registerAuthScheme(AuthState.PREEMPTIVE_AUTH_SCHEME, BearerAuthScheme.class.class);`

`client.getParams().setAuthenticationPreemptive(true);`

In the bearer example just above you will have noticed the trick we have to do to support preemptive authentication with bearer. 

With the line `AuthPolicy.registerAuthScheme(AuthState.PREEMPTIVE_AUTH_SCHEME, BearerAuthScheme.class);` **we are replacing the basic scheme properly registered in (1) with something like: "basic BearerScheme" which, at first sight, makes no sense.**

So, in this point, when we switch from an OAuth account to a basic one, the app gets the scheme "basic" but is still associated with the BearerScheme, triggering an error like:

`Credentials cannot be used for bearer authentication: org.apache.commons.httpclient.UsernamePasswordCredentials`

The opposite error also appears sometimes:

`Credentials cannot be used for basic authentication: com.owncloud.android.lib.common.authentication.oauth.BearerCredentials`

This is an ugly workaround we had to include and it doesn't seem to have an straightforward solution, so it might be a new reason to update our network library.
